File: client-js/client-v2/consumer/ConsumerManager.js
/**
 * Consumer manager for handling concurrent workers
 */

import * as logger from '../utils/logger.js'

export class ConsumerManager {
  #httpClient
  #queen

  constructor(httpClient, queen) {
    this.#httpClient = httpClient
    this.#queen = queen
  }

  async start(handler, options) {
    const {
      queue,
      partition,
      namespace,
      task,
      group,
      concurrency,
      batch,
      limit,
      idleMillis,
      autoAck,
      wait,
      timeoutMillis,
      renewLease,
      renewLeaseIntervalMillis,
      subscriptionMode,
      subscriptionFrom,
      each,
      signal
    } = options

    logger.log('ConsumerManager.start', { 
      queue, 
      partition, 
      namespace, 
      task, 
      group, 
      concurrency, 
      batch, 
      limit, 
      autoAck,
      wait,
      each
    })

    // Build the path and params for pop requests
    const path = this.#buildPath(queue, partition, namespace, task)
    const baseParams = this.#buildParams(batch, wait, timeoutMillis, group, subscriptionMode, subscriptionFrom, namespace, task, autoAck)

    // Use worker pool pattern if concurrency > 1 and each mode is enabled
    // This maintains constant concurrency by immediately spawning new workers
    if (concurrency > 1 && each) {
      logger.log('ConsumerManager.start', { mode: 'worker-pool', concurrency })
      await this.#workerPool(handler, path, baseParams, {
        concurrency,
        batch,
        limit,
        idleMillis,
        autoAck,
        wait,
        timeoutMillis,
        renewLease,
        renewLeaseIntervalMillis,
        signal,
        group
      })
    } else {
      // Use traditional parallel workers for batch mode or single worker
      logger.log('ConsumerManager.start', { mode: 'parallel-workers', concurrency })
      const workers = []
      for (let i = 0; i < concurrency; i++) {
        workers.push(this.#worker(i, handler, path, baseParams, {
          batch,
          limit,
          idleMillis,
          autoAck,
          wait,
          timeoutMillis,
          renewLease,
          renewLeaseIntervalMillis,
          each,
          signal,
          group
        }))
      }

      logger.log('ConsumerManager.start', { status: 'workers-started', count: concurrency })
      await Promise.all(workers)
    }
    
    logger.log('ConsumerManager.start', { status: 'completed' })
  }

  /**
   * Worker pool pattern - maintains constant concurrency
   * Continuously polls for messages and dispatches to worker pool
   */
  async #workerPool(handler, path, baseParams, options) {
    const {
      concurrency,
      batch,
      limit,
      idleMillis,
      autoAck,
      wait,
      timeoutMillis,
      renewLease,
      renewLeaseIntervalMillis,
      signal,
      group
    } = options

    // Message queue for pending work
    const messageQueue = []
    let processedCount = 0
    let activeWorkers = 0
    let polling = true
    let lastMessageTime = idleMillis ? Date.now() : null

    // Resolve when all workers complete
    let resolveComplete
    const completionPromise = new Promise(resolve => { resolveComplete = resolve })

    // Worker function that processes one message then exits
    const processOneMessage = async (message, workerId) => {
      activeWorkers++
      logger.log('ConsumerManager.workerPool', { workerId, status: 'processing', activeWorkers, queueSize: messageQueue.length })
      
      try {
        await this.#processMessage(message, handler, autoAck, group)
        processedCount++
        
        logger.log('ConsumerManager.workerPool', { workerId, status: 'completed', processedCount, activeWorkers })
      } catch (error) {
        logger.error('ConsumerManager.workerPool', { workerId, error: error.message })
      } finally {
        activeWorkers--
        
        // Check if we should stop
        if (limit && processedCount >= limit) {
          polling = false
        }
        
        // Try to spawn another worker if there's work available
        if (messageQueue.length > 0 && polling) {
          const nextMessage = messageQueue.shift()
          if (nextMessage) {
            // Spawn new worker immediately (don't await)
            processOneMessage(nextMessage, workerId).catch(err => {
              logger.error('ConsumerManager.workerPool', { workerId, error: err.message })
            })
          }
        }
        
        // If no more work and not polling, signal completion
        if (!polling && activeWorkers === 0 && messageQueue.length === 0) {
          resolveComplete()
        }
      }
    }

    // Poller function - continuously fetches messages
    const poller = async () => {
      let nextWorkerId = 0
      
      while (polling) {
        // Check abort signal
        if (signal && signal.aborted) {
          logger.log('ConsumerManager.workerPool', { status: 'aborted', processedCount })
          polling = false
          break
        }

        // Check limit
        if (limit && processedCount >= limit) {
          logger.log('ConsumerManager.workerPool', { status: 'limit-reached', processedCount, limit })
          polling = false
          break
        }

        // Check idle timeout
        if (idleMillis && lastMessageTime) {
          const idleTime = Date.now() - lastMessageTime
          if (idleTime >= idleMillis) {
            logger.log('ConsumerManager.workerPool', { status: 'idle-timeout', processedCount, idleTime })
            polling = false
            break
          }
        }

        try {
          // Poll for messages
          const clientTimeout = wait ? timeoutMillis + 5000 : timeoutMillis
          const result = await this.#httpClient.get(`${path}?${baseParams}`, clientTimeout)

          // Handle empty response
          if (!result || !result.messages || result.messages.length === 0) {
            if (wait) {
              continue // Long polling timeout, retry
            } else {
              await new Promise(resolve => setTimeout(resolve, 100))
              continue
            }
          }

          const messages = result.messages.filter(msg => msg != null)
          if (messages.length === 0) {
            continue
          }

          logger.log('ConsumerManager.workerPool', { status: 'messages-received', count: messages.length, queueSize: messageQueue.length, activeWorkers })

          // Update last message time
          if (idleMillis) {
            lastMessageTime = Date.now()
          }

          // Set up lease renewal for all messages if enabled
          if (renewLease && renewLeaseIntervalMillis) {
            const renewalTimer = this.#setupLeaseRenewal(messages, renewLeaseIntervalMillis)
            // Clear timer after processing window
            setTimeout(() => clearInterval(renewalTimer), renewLeaseIntervalMillis * 3)
          }

          // Dispatch messages to worker pool
          for (const message of messages) {
            if (!polling) break
            if (limit && processedCount >= limit) {
              polling = false
              break
            }

            // If we have available worker slots, spawn immediately
            if (activeWorkers < concurrency) {
              processOneMessage(message, nextWorkerId++).catch(err => {
                logger.error('ConsumerManager.workerPool', { error: err.message })
              })
            } else {
              // Otherwise queue for later
              messageQueue.push(message)
            }
          }

        } catch (error) {
          // Check if this is a timeout error (expected for long polling)
          const isTimeoutError = error.name === 'AbortError' ||
                                error.message?.includes('timeout')

          if (isTimeoutError && wait) {
            continue
          }

          // Check if network error
          const isNetworkError = error.message?.includes('fetch failed') ||
                                error.message?.includes('ECONNREFUSED') ||
                                error.code === 'ECONNREFUSED'

          if (isNetworkError) {
            logger.warn('ConsumerManager.workerPool', { error: 'network', message: error.message })
            console.warn(`Worker pool: Network error - ${error.message}`)
            await new Promise(resolve => setTimeout(resolve, 1000))
            continue
          }

          // Other errors - log and stop
          logger.error('ConsumerManager.workerPool', { error: error.message })
          polling = false
          break
        }
      }

      // Wait for all active workers to complete
      logger.log('ConsumerManager.workerPool', { status: 'polling-stopped', activeWorkers, queueSize: messageQueue.length })
    }

    // Start poller
    poller().catch(err => {
      logger.error('ConsumerManager.workerPool', { error: err.message })
      polling = false
    })

    // Wait for completion
    await completionPromise
    
    logger.log('ConsumerManager.workerPool', { status: 'completed', processedCount })
  }

  async #worker(workerId, handler, path, baseParams, options) {
    const {
      batch,
      limit,
      idleMillis,
      autoAck,
      wait,
      timeoutMillis,
      renewLease,
      renewLeaseIntervalMillis,
      each,
      signal,
      group
    } = options

    logger.log('ConsumerManager.worker', { workerId, status: 'started', limit, idleMillis })
    
    let processedCount = 0
    let lastMessageTime = idleMillis ? Date.now() : null

    while (true) {
      // Check abort signal
      if (signal && signal.aborted) {
        logger.log('ConsumerManager.worker', { workerId, status: 'aborted', processedCount })
        break
      }

      // Check limit
      if (limit && processedCount >= limit) {
        logger.log('ConsumerManager.worker', { workerId, status: 'limit-reached', processedCount, limit })
        break
      }

      // Check idle timeout
      if (idleMillis && lastMessageTime) {
        const idleTime = Date.now() - lastMessageTime
        if (idleTime >= idleMillis) {
          logger.log('ConsumerManager.worker', { workerId, status: 'idle-timeout', processedCount, idleTime })
          break
        }
      }

      try {
        // Pop messages
        const clientTimeout = wait ? timeoutMillis + 5000 : timeoutMillis
        const result = await this.#httpClient.get(`${path}?${baseParams}`, clientTimeout)

        // Handle empty response
        if (!result || !result.messages || result.messages.length === 0) {
          if (wait) {
            continue // Long polling timeout, retry
          } else {
            // Short delay before retry
            await new Promise(resolve => setTimeout(resolve, 100))
            continue
          }
        }

        const messages = result.messages.filter(msg => msg != null)

        if (messages.length === 0) {
          continue
        }

        logger.log('ConsumerManager.worker', { workerId, status: 'messages-received', count: messages.length })

        // Update last message time
        if (idleMillis) {
          lastMessageTime = Date.now()
        }

        // Set up lease renewal if enabled
        let renewalTimer = null
        if (renewLease && renewLeaseIntervalMillis) {
          renewalTimer = this.#setupLeaseRenewal(messages, renewLeaseIntervalMillis)
        }

        try {
          // Process messages
          if (each) {
            // Process one at a time
            for (const message of messages) {
              if (signal && signal.aborted) break

              await this.#processMessage(message, handler, autoAck, group)
              processedCount++

              if (limit && processedCount >= limit) break
            }
          } else {
            // Process as batch
            await this.#processBatch(messages, handler, autoAck, group)
            processedCount += messages.length
          }
          
          logger.log('ConsumerManager.worker', { workerId, status: 'messages-processed', count: messages.length, total: processedCount })
        } finally {
          // Clear renewal timer
          if (renewalTimer) {
            clearInterval(renewalTimer)
          }
        }

      } catch (error) {
        // Check if this is a timeout error (expected for long polling)
        const isTimeoutError = error.name === 'AbortError' ||
                              error.message?.includes('timeout')

        if (isTimeoutError && wait) {
          continue // Retry on timeout
        }

        // Check if network error
        const isNetworkError = error.message?.includes('fetch failed') ||
                              error.message?.includes('ECONNREFUSED') ||
                              error.code === 'ECONNREFUSED'

        if (isNetworkError) {
          logger.warn('ConsumerManager.worker', { workerId, error: 'network', message: error.message })
          console.warn(`Worker ${workerId}: Network error - ${error.message}`)
          // Wait before retry
          await new Promise(resolve => setTimeout(resolve, 1000))
          continue
        }

        // Other errors - rethrow
        logger.error('ConsumerManager.worker', { workerId, error: error.message })
        throw error
      }
    }
    
    logger.log('ConsumerManager.worker', { workerId, status: 'stopped', processedCount })
  }

  async #processMessage(message, handler, autoAck, group) {
    try {
      await handler(message)

      // Auto-ack on success if enabled
      if (autoAck) {
        const context = group ? { group } : {}
        await this.#queen.ack(message, true, context)
        logger.log('ConsumerManager.processMessage', { transactionId: message.transactionId, status: 'acked' })
      }
    } catch (error) {
      // Auto-nack on error if enabled
      if (autoAck) {
        const context = group ? { group } : {}
        await this.#queen.ack(message, false, context)
        logger.error('ConsumerManager.processMessage', { transactionId: message.transactionId, error: error.message, status: 'nacked' })
        // Don't rethrow when autoAck is enabled - NACK was already sent
        // This allows the consumer to continue and retry
        return
      }
      logger.error('ConsumerManager.processMessage', { transactionId: message.transactionId, error: error.message })
      throw error
    }
  }

  async #processBatch(messages, handler, autoAck, group) {
    try {
      await handler(messages)

      // Auto-ack on success if enabled
      if (autoAck) {
        const context = group ? { group } : {}
        await this.#queen.ack(messages, true, context)
        logger.log('ConsumerManager.processBatch', { count: messages.length, status: 'acked' })
      }
    } catch (error) {
      // Auto-nack on error if enabled
      if (autoAck) {
        const context = group ? { group } : {}
        await this.#queen.ack(messages, false, context)
        logger.error('ConsumerManager.processBatch', { count: messages.length, error: error.message, status: 'nacked' })
        // Don't rethrow when autoAck is enabled - NACK was already sent
        // This allows the consumer to continue and retry
        return
      }
      logger.error('ConsumerManager.processBatch', { count: messages.length, error: error.message })
      throw error
    }
  }

  #setupLeaseRenewal(messages, intervalMillis) {
    const leaseIds = messages.map(m => m.leaseId).filter(id => id != null)

    if (leaseIds.length === 0) return null

    return setInterval(async () => {
      try {
        await this.#queen.renew(messages)
      } catch (error) {
        console.error('Lease renewal failed:', error)
      }
    }, intervalMillis)
  }

  #buildPath(queue, partition, namespace, task) {
    if (queue) {
      if (partition) {
        return `/api/v1/pop/queue/${queue}/partition/${partition}`
      }
      return `/api/v1/pop/queue/${queue}`
    }

    if (namespace || task) {
      return '/api/v1/pop'
    }

    throw new Error('Must specify queue, namespace, or task')
  }

  #buildParams(batch, wait, timeoutMillis, group, subscriptionMode, subscriptionFrom, namespace, task, autoAck) {
    const params = new URLSearchParams({
      batch: batch.toString(),
      wait: wait.toString(),
      timeout: timeoutMillis.toString()  // Server expects 'timeout', not 'timeoutMillis'
    })

    if (group) params.append('consumerGroup', group)
    if (subscriptionMode) params.append('subscriptionMode', subscriptionMode)
    if (subscriptionFrom) params.append('subscriptionFrom', subscriptionFrom)
    if (namespace) params.append('namespace', namespace)
    if (task) params.append('task', task)
    // NEVER send autoAck for consume - client always manages acking
    // autoAck is only for pop() where server auto-acks immediately

    return params
  }
}

File: client-go/consumer_manager.go
package queen

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"sync"
	"time"
)

// ConsumerManager manages concurrent consumer workers.
type ConsumerManager struct {
	httpClient *HttpClient
	queen      *Queen
}

// NewConsumerManager creates a new ConsumerManager.
func NewConsumerManager(httpClient *HttpClient, queen *Queen) *ConsumerManager {
	return &ConsumerManager{
		httpClient: httpClient,
		queen:      queen,
	}
}

// Start starts consumer workers for single message handling.
func (cm *ConsumerManager) Start(ctx context.Context, handler MessageHandler, opts ConsumeOptions) error {
	return cm.startWorkers(ctx, opts, func(ctx context.Context, msgs []*Message) error {
		for _, msg := range msgs {
			if err := handler(ctx, msg); err != nil {
				return err
			}
		}
		return nil
	}, false)
}

// StartBatch starts consumer workers for batch message handling.
func (cm *ConsumerManager) StartBatch(ctx context.Context, handler BatchMessageHandler, opts ConsumeOptions) error {
	return cm.startWorkers(ctx, opts, handler, true)
}

// startWorkers starts the consumer workers.
func (cm *ConsumerManager) startWorkers(ctx context.Context, opts ConsumeOptions, handler BatchMessageHandler, isBatch bool) error {
	// Build path and params
	path := cm.buildPath(opts)
	baseParams := cm.buildParams(opts)

	// Generate affinity key for consistent routing
	affinityKey := cm.getAffinityKey(opts)

	logInfo("ConsumerManager.Start", map[string]interface{}{
		"queue":       opts.Queue,
		"partition":   opts.Partition,
		"namespace":   opts.Namespace,
		"task":        opts.Task,
		"group":       opts.Group,
		"concurrency": opts.Concurrency,
		"batch":       opts.Batch,
		"limit":       opts.Limit,
		"autoAck":     opts.AutoAck,
		"wait":        opts.Wait,
		"each":        opts.Each,
	})

	// Start workers
	var wg sync.WaitGroup
	errChan := make(chan error, opts.Concurrency)

	for i := 0; i < opts.Concurrency; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			err := cm.worker(ctx, workerID, handler, isBatch, path, baseParams, affinityKey, opts)
			if err != nil && err != context.Canceled {
				errChan <- err
			}
		}(i)
	}

	// Wait for all workers to complete
	wg.Wait()
	close(errChan)

	// Return first error if any
	for err := range errChan {
		return err
	}

	logInfo("ConsumerManager.Start", map[string]interface{}{
		"status": "completed",
	})

	return nil
}

// worker is the main worker loop.
func (cm *ConsumerManager) worker(
	ctx context.Context,
	workerID int,
	handler BatchMessageHandler,
	isBatch bool,
	path string,
	baseParams string,
	affinityKey string,
	opts ConsumeOptions,
) error {
	logDebug("ConsumerManager.worker", map[string]interface{}{
		"workerId": workerID,
		"status":   "started",
		"limit":    opts.Limit,
		"idleMs":   opts.IdleMillis,
	})

	processedCount := 0
	var lastMessageTime time.Time
	if opts.IdleMillis > 0 {
		lastMessageTime = time.Now()
	}

	for {
		// Check context cancellation
		select {
		case <-ctx.Done():
			logDebug("ConsumerManager.worker", map[string]interface{}{
				"workerId":       workerID,
				"status":         "cancelled",
				"processedCount": processedCount,
			})
			return ctx.Err()
		default:
		}

		// Check limit
		if opts.Limit > 0 && processedCount >= opts.Limit {
			logDebug("ConsumerManager.worker", map[string]interface{}{
				"workerId":       workerID,
				"status":         "limit-reached",
				"processedCount": processedCount,
				"limit":          opts.Limit,
			})
			return nil
		}

		// Check idle timeout
		if opts.IdleMillis > 0 && !lastMessageTime.IsZero() {
			idleTime := time.Since(lastMessageTime).Milliseconds()
			if idleTime >= int64(opts.IdleMillis) {
				logDebug("ConsumerManager.worker", map[string]interface{}{
					"workerId":       workerID,
					"status":         "idle-timeout",
					"processedCount": processedCount,
					"idleTime":       idleTime,
				})
				return nil
			}
		}

		// Pop messages
		clientTimeout := opts.TimeoutMillis
		if opts.Wait {
			clientTimeout += 5000 // Add 5s buffer for long polling
		}

		fullPath := path
		if baseParams != "" {
			fullPath += "?" + baseParams
		}

		result, err := cm.httpClient.Get(ctx, fullPath, clientTimeout, affinityKey)
		if err != nil {
			// Check if context was cancelled
			if ctx.Err() != nil {
				return ctx.Err()
			}

			// Check if this is a timeout error (expected for long polling)
			if isTimeoutError(err) && opts.Wait {
				continue // Retry on timeout
			}

			// Network error - wait and retry
			if isNetworkError(err) {
				logWarn("ConsumerManager.worker", map[string]interface{}{
					"workerId": workerID,
					"error":    "network",
					"message":  err.Error(),
				})
				select {
				case <-ctx.Done():
					return ctx.Err()
				case <-time.After(time.Second):
					continue
				}
			}

			// Other errors - log and continue
			logError("ConsumerManager.worker", map[string]interface{}{
				"workerId": workerID,
				"error":    err.Error(),
			})
			continue
		}

		// Parse messages
		messages := parseMessages(result)
		if len(messages) == 0 {
			if opts.Wait {
				continue // Long polling timeout, retry
			}
			// Short delay before retry
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(100 * time.Millisecond):
				continue
			}
		}

		logDebug("ConsumerManager.worker", map[string]interface{}{
			"workerId": workerID,
			"status":   "messages-received",
			"count":    len(messages),
		})

		// Update last message time
		if opts.IdleMillis > 0 {
			lastMessageTime = time.Now()
		}

		// Enhance messages with trace method
		cm.enhanceMessagesWithTrace(messages, opts.Group)

		// Set up lease renewal if enabled
		var renewalCancel context.CancelFunc
		if opts.RenewLease && opts.RenewLeaseIntervalMillis > 0 {
			var renewalCtx context.Context
			renewalCtx, renewalCancel = context.WithCancel(ctx)
			go cm.leaseRenewalLoop(renewalCtx, messages, opts.RenewLeaseIntervalMillis)
		}

		// Process messages
		var processErr error
		if opts.Each {
			// Process one at a time
			for _, msg := range messages {
				select {
				case <-ctx.Done():
					if renewalCancel != nil {
						renewalCancel()
					}
					return ctx.Err()
				default:
				}

				processErr = cm.processMessage(ctx, msg, handler, opts.AutoAck, opts.Group)
				processedCount++

				if opts.Limit > 0 && processedCount >= opts.Limit {
					break
				}
			}
		} else {
			// Process as batch (or single message if batch=1)
			if opts.Batch == 1 && len(messages) == 1 {
				// For batch=1, pass single message (not array) - wrap in single-element batch
				processErr = cm.processMessage(ctx, messages[0], handler, opts.AutoAck, opts.Group)
				processedCount++
			} else {
				// For batch>1, pass array of messages
				processErr = cm.processBatch(ctx, messages, handler, opts.AutoAck, opts.Group)
				processedCount += len(messages)
			}
		}

		// Cancel renewal
		if renewalCancel != nil {
			renewalCancel()
		}

		if processErr != nil && !opts.AutoAck {
			// If auto-ack is disabled and handler failed, propagate error
			return processErr
		}

		logDebug("ConsumerManager.worker", map[string]interface{}{
			"workerId":       workerID,
			"status":         "messages-processed",
			"count":          len(messages),
			"totalProcessed": processedCount,
		})
	}
}

// processMessage processes a single message.
func (cm *ConsumerManager) processMessage(ctx context.Context, msg *Message, handler BatchMessageHandler, autoAck bool, group string) error {
	err := handler(ctx, []*Message{msg})

	if autoAck {
		ackOpts := AckOptions{}
		if group != "" {
			ackOpts.ConsumerGroup = group
		}

		if err != nil {
			// Auto-nack on error
			ackOpts.Error = err.Error()
			_, ackErr := cm.queen.Ack(ctx, msg, false, ackOpts)
			if ackErr != nil {
				logError("ConsumerManager.processMessage", map[string]interface{}{
					"transactionId": msg.TransactionID,
					"error":         ackErr.Error(),
					"status":        "nack-failed",
				})
			} else {
				logDebug("ConsumerManager.processMessage", map[string]interface{}{
					"transactionId": msg.TransactionID,
					"status":        "nacked",
				})
			}
			// Don't propagate error when autoAck is enabled
			return nil
		}

		// Auto-ack on success
		_, ackErr := cm.queen.Ack(ctx, msg, true, ackOpts)
		if ackErr != nil {
			logError("ConsumerManager.processMessage", map[string]interface{}{
				"transactionId": msg.TransactionID,
				"error":         ackErr.Error(),
				"status":        "ack-failed",
			})
		} else {
			logDebug("ConsumerManager.processMessage", map[string]interface{}{
				"transactionId": msg.TransactionID,
				"status":        "acked",
			})
		}
		return nil
	}

	return err
}

// processBatch processes a batch of messages.
func (cm *ConsumerManager) processBatch(ctx context.Context, msgs []*Message, handler BatchMessageHandler, autoAck bool, group string) error {
	err := handler(ctx, msgs)

	if autoAck {
		ackOpts := AckOptions{}
		if group != "" {
			ackOpts.ConsumerGroup = group
		}

		if err != nil {
			// Auto-nack on error
			ackOpts.Error = err.Error()
			_, ackErr := cm.queen.Ack(ctx, msgs, false, ackOpts)
			if ackErr != nil {
				logError("ConsumerManager.processBatch", map[string]interface{}{
					"count":  len(msgs),
					"error":  ackErr.Error(),
					"status": "nack-failed",
				})
			} else {
				logDebug("ConsumerManager.processBatch", map[string]interface{}{
					"count":  len(msgs),
					"status": "nacked",
				})
			}
			return nil
		}

		// Auto-ack on success
		_, ackErr := cm.queen.Ack(ctx, msgs, true, ackOpts)
		if ackErr != nil {
			logError("ConsumerManager.processBatch", map[string]interface{}{
				"count":  len(msgs),
				"error":  ackErr.Error(),
				"status": "ack-failed",
			})
		} else {
			logDebug("ConsumerManager.processBatch", map[string]interface{}{
				"count":  len(msgs),
				"status": "acked",
			})
		}
		return nil
	}

	return err
}

// leaseRenewalLoop renews leases periodically.
func (cm *ConsumerManager) leaseRenewalLoop(ctx context.Context, messages []*Message, intervalMillis int) {
	ticker := time.NewTicker(time.Duration(intervalMillis) * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			_, err := cm.queen.Renew(ctx, messages)
			if err != nil {
				logWarn("ConsumerManager.leaseRenewal", map[string]interface{}{
					"error": err.Error(),
				})
			}
		}
	}
}

// enhanceMessagesWithTrace adds the trace method to messages.
func (cm *ConsumerManager) enhanceMessagesWithTrace(messages []*Message, group string) {
	consumerGroup := group
	if consumerGroup == "" {
		consumerGroup = QueueModeConsumerGroup
	}

	for _, msg := range messages {
		// Create closure for this message
		msgCopy := msg
		groupCopy := consumerGroup

		msg.SetTrace(func(ctx context.Context, config TraceConfig) (*TraceResponse, error) {
			// CRITICAL: NEVER CRASH - just log and return gracefully
			defer func() {
				if r := recover(); r != nil {
					logError("ConsumerManager.trace", map[string]interface{}{
						"transactionId": msgCopy.TransactionID,
						"error":         fmt.Sprintf("panic: %v", r),
					})
				}
			}()

			// Validate required structure
			if config.Data == nil {
				logWarn("ConsumerManager.trace", map[string]interface{}{
					"error":         "Invalid trace config: requires data field",
					"transactionId": msgCopy.TransactionID,
				})
				return &TraceResponse{
					Success: false,
					Error:   "Invalid trace config: requires data field",
				}, nil
			}

			// Normalize trace names
			var traceNames []string
			if config.TraceName != "" {
				traceNames = []string{config.TraceName}
			} else if len(config.TraceNames) > 0 {
				traceNames = config.TraceNames
			}

			// Build request
			req := traceRequest{
				TransactionID: msgCopy.TransactionID,
				PartitionID:   msgCopy.PartitionID,
				ConsumerGroup: groupCopy,
				TraceNames:    traceNames,
				EventType:     config.EventType,
				Data:          config.Data,
			}

			if req.EventType == "" {
				req.EventType = "info"
			}

			// Make request
			_, err := cm.httpClient.Post(ctx, "/api/v1/traces", req)
			if err != nil {
				logError("ConsumerManager.trace", map[string]interface{}{
					"transactionId": msgCopy.TransactionID,
					"error":         err.Error(),
				})
				return &TraceResponse{
					Success: false,
					Error:   err.Error(),
				}, nil
			}

			logDebug("ConsumerManager.trace", map[string]interface{}{
				"transactionId": msgCopy.TransactionID,
				"success":       true,
				"traceNames":    traceNames,
			})

			return &TraceResponse{Success: true}, nil
		})
	}
}

// buildPath builds the pop path.
func (cm *ConsumerManager) buildPath(opts ConsumeOptions) string {
	if opts.Queue != "" {
		if opts.Partition != "" && opts.Partition != DefaultPartition {
			return fmt.Sprintf("/api/v1/pop/queue/%s/partition/%s",
				url.PathEscape(opts.Queue), url.PathEscape(opts.Partition))
		}
		return fmt.Sprintf("/api/v1/pop/queue/%s", url.PathEscape(opts.Queue))
	}

	if opts.Namespace != "" || opts.Task != "" {
		return "/api/v1/pop"
	}

	return "/api/v1/pop"
}

// buildParams builds the query parameters.
func (cm *ConsumerManager) buildParams(opts ConsumeOptions) string {
	params := url.Values{}

	params.Set("batch", strconv.Itoa(opts.Batch))
	params.Set("wait", strconv.FormatBool(opts.Wait))
	params.Set("timeout", strconv.Itoa(opts.TimeoutMillis))

	if opts.Group != "" {
		params.Set("consumerGroup", opts.Group)
	}
	if opts.SubscriptionMode != "" {
		params.Set("subscriptionMode", opts.SubscriptionMode)
	}
	if opts.SubscriptionFrom != "" {
		params.Set("subscriptionFrom", opts.SubscriptionFrom)
	}
	if opts.Namespace != "" {
		params.Set("namespace", opts.Namespace)
	}
	if opts.Task != "" {
		params.Set("task", opts.Task)
	}
	// NEVER send autoAck for consume - client always manages acking

	return params.Encode()
}

// getAffinityKey generates the affinity key for consistent routing.
func (cm *ConsumerManager) getAffinityKey(opts ConsumeOptions) string {
	if opts.Queue != "" {
		part := opts.Partition
		if part == "" {
			part = "*"
		}
		grp := opts.Group
		if grp == "" {
			grp = QueueModeConsumerGroup
		}
		return fmt.Sprintf("%s:%s:%s", opts.Queue, part, grp)
	}

	if opts.Namespace != "" || opts.Task != "" {
		ns := opts.Namespace
		if ns == "" {
			ns = "*"
		}
		task := opts.Task
		if task == "" {
			task = "*"
		}
		grp := opts.Group
		if grp == "" {
			grp = QueueModeConsumerGroup
		}
		return fmt.Sprintf("%s:%s:%s", ns, task, grp)
	}

	return ""
}
